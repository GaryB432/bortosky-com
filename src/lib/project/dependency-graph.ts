import type {
  EdgeDataDefinition,
  EdgeDefinition,
  ElementsDefinition,
  NodeDataDefinition,
  NodeDefinition,
} from "cytoscape";
import type { GaryProject, NxProjectJson, PackageJson } from "./project";

type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };

type NodeDataWithId = WithRequired<NodeDataDefinition, "id" | "label">;
type EdgeDataWithId = WithRequired<EdgeDataDefinition, "id">;

type DepType = "dev" | "runtime";

export class DependencyGraph {
  private mes = new Map<string, EdgeDefinition>();
  private mns = new Map<string, NodeDefinition>();

  public digestWorkspace(ws: GaryProject): void {
    const wsNode = this.getWorkspaceData(ws);
    this.mns.set(wsNode.id, { data: wsNode, classes: ["ws"] });

    this.digestProject(ws.root, undefined);

    for (const subp of ws.projects) {
      this.digestProject(subp, ws);
    }
  }

  public elements(): ElementsDefinition {
    const nodes: NodeDefinition[] = Array.from(this.mns.values());
    const edges: EdgeDefinition[] = Array.from(this.mes.values());
    return { nodes, edges };
  }

  private digestDependencyRecord(
    dependencyRecord: Record<string, string> | undefined,
    dependerNodeData: NodeDataWithId,
    depType: DepType,
  ): void {
    for (const [k, v] of Object.entries(dependencyRecord ?? {})) {
      const pkgNodeData = this.getPackageData(k);
      const versionNodeData = this.getPackageVersionData(k, v);

      this.mns.set(pkgNodeData.id, { data: pkgNodeData, classes: [] });
      this.mns.set(versionNodeData.id, {
        data: versionNodeData,
        classes: [],
      });

      const pvEdgeData = this.edgeForPackageVersionPackage(
        pkgNodeData,
        versionNodeData,
      );
      this.mes.set(pvEdgeData.id, { data: pvEdgeData });

      const pppvEdgeData = this.edgeForProjectPackageVersion(
        versionNodeData,
        dependerNodeData,
      );

      this.mes.set(pppvEdgeData.id, {
        data: pppvEdgeData,
        classes: ["dependency", depType],
      });
    }
  }

  private digestProject(
    depender: PackageJson | NxProjectJson,
    ws: GaryProject | undefined,
  ): void {
    const dep = depender as PackageJson;
    if (dep.name && (dep.dependencies || dep.devDependencies)) {
      const workspace: GaryProject = ws ?? {
        root: { name: dep.name, version: "ephemeral" },
        projects: [],
      };
      const wsNodeData = this.getWorkspaceData(workspace);
      const dependerNodeData = this.getProjectData(dep, wsNodeData);
      this.mns.set(dependerNodeData.id, {
        data: dependerNodeData,
        classes: ws ? ["subp"] : ["subp", "root"],
      });
      const wsProjEdgeData = this.edgeForWorkspaceProject(
        dependerNodeData,
        wsNodeData,
      );
      this.mes.set(wsProjEdgeData.id, {
        data: wsProjEdgeData,
      });

      this.digestDependencyRecord(dep.devDependencies, dependerNodeData, "dev");
      this.digestDependencyRecord(
        dep.dependencies,
        dependerNodeData,
        "runtime",
      );
    }
  }

  private edgeForPackageVersionPackage(
    a: NodeDataWithId,
    b: NodeDataWithId,
  ): EdgeDataWithId {
    const id = a.id.concat(">").concat(b.id);
    return { id, source: a.id, target: b.id };
  }

  private edgeForProjectPackageVersion(
    a: NodeDataWithId,
    b: NodeDataWithId,
  ): EdgeDataWithId {
    const id = a.id.concat(">").concat(b.id);
    return { id, source: a.id, target: b.id };
  }

  private edgeForWorkspaceProject(
    a: NodeDataWithId,
    b: NodeDataWithId,
  ): EdgeDataWithId {
    const id = a.id.concat(">").concat(b.id);
    return { id, source: a.id, target: b.id };
  }

  private getPackageData(pkg: string): NodeDataWithId {
    const id = pkg.concat("@npm");
    const label = pkg;

    return { id, label };
  }

  private getPackageVersionData(pkg: string, version: string): NodeDataWithId {
    const id = pkg.concat("@npm").concat("@").concat(version);
    const label = version;

    return { id, label };
  }

  private getProjectData(
    pj: PackageJson | NxProjectJson,
    wsNode: NodeDataWithId,
  ): NodeDataWithId {
    const name = pj.name ?? "root";
    const id = wsNode.id.concat("@").concat(name);
    const label = name;

    return { id, label };
  }

  private getWorkspaceData(ws: GaryProject): NodeDataWithId {
    const id = ws.root.name;
    const label = ws.root.name;

    return { id, label };
  }
}
