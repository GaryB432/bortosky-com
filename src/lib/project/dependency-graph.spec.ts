import { beforeEach, describe, expect, test } from "vitest";
import { DependencyGraph } from "./dependency-graph";
import type { GaryProject } from "./project";

/**
 * jest | ^07.0.0 | workspace-a | workspace-a@ws-a | dev
 *                | workspace-a | workspace-a@ws-b | dev
 */

describe("DependencyGraph", () => {
  let dependencyGraph: DependencyGraph;
  beforeEach(() => {
    dependencyGraph = new DependencyGraph();
  });

  test("digestWorkspace", () => {
    const asdf = dependencyGraph.digestWorkspace(subject[0]);
    expect(dependencyGraph.elements()).toMatchSnapshot();
  });
});

const subject: GaryProject[] = [
  {
    root: {
      name: "workspace-a",
      version: "whatever",
      devDependencies: { jest: "3.0.0" },
      dependencies: { album: "4.0.0" },
    },
    projects: [
      {
        name: "wsa-a",
        version: "",
        type: "module",
        devDependencies: {
          desk: "^1.4.9",
          window: "5.4.2",
          "eslint-plugin-gb": "skipped",
        },
        dependencies: { throttle: "^4.1.2", brake: "~1.4.6" },
      },
    ],
  },
  {
    root: {
      name: "workspace-b",
      version: "0",
      devDependencies: { jest: "^4.0.0" },
      dependencies: { album: "2" },
    },
    projects: [],
  },
];
