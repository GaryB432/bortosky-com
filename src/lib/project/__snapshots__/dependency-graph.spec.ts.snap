// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DependencyGraph > digestWorkspace 1`] = `
{
  "edges": [
    {
      "data": {
        "id": "workspace-a@workspace-a>workspace-a",
        "source": "workspace-a@workspace-a",
        "target": "workspace-a",
      },
    },
    {
      "data": {
        "id": "jest@npm>jest@npm@3.0.0",
        "source": "jest@npm",
        "target": "jest@npm@3.0.0",
      },
    },
    {
      "classes": [
        "dependency",
        "dev",
      ],
      "data": {
        "id": "jest@npm@3.0.0>workspace-a@workspace-a",
        "source": "jest@npm@3.0.0",
        "target": "workspace-a@workspace-a",
      },
    },
    {
      "data": {
        "id": "album@npm>album@npm@4.0.0",
        "source": "album@npm",
        "target": "album@npm@4.0.0",
      },
    },
    {
      "classes": [
        "dependency",
        "runtime",
      ],
      "data": {
        "id": "album@npm@4.0.0>workspace-a@workspace-a",
        "source": "album@npm@4.0.0",
        "target": "workspace-a@workspace-a",
      },
    },
    {
      "data": {
        "id": "workspace-a@wsa-a>workspace-a",
        "source": "workspace-a@wsa-a",
        "target": "workspace-a",
      },
    },
    {
      "data": {
        "id": "desk@npm>desk@npm@^1.4.9",
        "source": "desk@npm",
        "target": "desk@npm@^1.4.9",
      },
    },
    {
      "classes": [
        "dependency",
        "dev",
      ],
      "data": {
        "id": "desk@npm@^1.4.9>workspace-a@wsa-a",
        "source": "desk@npm@^1.4.9",
        "target": "workspace-a@wsa-a",
      },
    },
    {
      "data": {
        "id": "window@npm>window@npm@5.4.2",
        "source": "window@npm",
        "target": "window@npm@5.4.2",
      },
    },
    {
      "classes": [
        "dependency",
        "dev",
      ],
      "data": {
        "id": "window@npm@5.4.2>workspace-a@wsa-a",
        "source": "window@npm@5.4.2",
        "target": "workspace-a@wsa-a",
      },
    },
    {
      "data": {
        "id": "eslint-plugin-gb@npm>eslint-plugin-gb@npm@skipped",
        "source": "eslint-plugin-gb@npm",
        "target": "eslint-plugin-gb@npm@skipped",
      },
    },
    {
      "classes": [
        "dependency",
        "dev",
      ],
      "data": {
        "id": "eslint-plugin-gb@npm@skipped>workspace-a@wsa-a",
        "source": "eslint-plugin-gb@npm@skipped",
        "target": "workspace-a@wsa-a",
      },
    },
    {
      "data": {
        "id": "throttle@npm>throttle@npm@^4.1.2",
        "source": "throttle@npm",
        "target": "throttle@npm@^4.1.2",
      },
    },
    {
      "classes": [
        "dependency",
        "runtime",
      ],
      "data": {
        "id": "throttle@npm@^4.1.2>workspace-a@wsa-a",
        "source": "throttle@npm@^4.1.2",
        "target": "workspace-a@wsa-a",
      },
    },
    {
      "data": {
        "id": "brake@npm>brake@npm@~1.4.6",
        "source": "brake@npm",
        "target": "brake@npm@~1.4.6",
      },
    },
    {
      "classes": [
        "dependency",
        "runtime",
      ],
      "data": {
        "id": "brake@npm@~1.4.6>workspace-a@wsa-a",
        "source": "brake@npm@~1.4.6",
        "target": "workspace-a@wsa-a",
      },
    },
  ],
  "nodes": [
    {
      "classes": [
        "ws",
      ],
      "data": {
        "id": "workspace-a",
        "label": "workspace-a",
      },
    },
    {
      "classes": [
        "subp",
        "root",
      ],
      "data": {
        "id": "workspace-a@workspace-a",
        "label": "workspace-a",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "jest@npm",
        "label": "jest",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "jest@npm@3.0.0",
        "label": "3.0.0",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "album@npm",
        "label": "album",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "album@npm@4.0.0",
        "label": "4.0.0",
      },
    },
    {
      "classes": [
        "subp",
      ],
      "data": {
        "id": "workspace-a@wsa-a",
        "label": "wsa-a",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "desk@npm",
        "label": "desk",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "desk@npm@^1.4.9",
        "label": "^1.4.9",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "window@npm",
        "label": "window",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "window@npm@5.4.2",
        "label": "5.4.2",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "eslint-plugin-gb@npm",
        "label": "eslint-plugin-gb",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "eslint-plugin-gb@npm@skipped",
        "label": "skipped",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "throttle@npm",
        "label": "throttle",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "throttle@npm@^4.1.2",
        "label": "^4.1.2",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "brake@npm",
        "label": "brake",
      },
    },
    {
      "classes": [],
      "data": {
        "id": "brake@npm@~1.4.6",
        "label": "~1.4.6",
      },
    },
  ],
}
`;
